Jasmine/Tetris/ViewModels/TetrisGameViewModel.swift

```diff
@@ -1,4 +1,158 @@
+                continue
+            }
+
+            for i in validPhraseRange {
+                let coordinate = searchByRow ? Coordinate(row: index, col: i) : Coordinate(row: i, col: index)
+                matchedCoordinates.insert(coordinate)
+            }
+            return matchedCoordinates
+        }
+        return nil
+    }
+
+    /// Gets the tile text at coordinate specified by `row` and `col`
+    /// returns " " if no tile is present so that phrases separated by gaps don't get matched
+    private func getTileText(row: Int, col: Int) -> String {
+        return tetrisGrid.get(at: Coordinate(row: row, col: col)) ?? " "
+    }
+
+    /// Shifts all the tiles above `coordinates` 1 row down.
+    /// Starts from the row right above the coordinates so that it can break once an empty tile is encountered
+    fileprivate func shiftDownTiles(_ coordinates: Set<Coordinate>) {
+        var coordinatesToShift: [(from: Coordinate, to: Coordinate)] = []
+        for coordinate in coordinates {
+            for row in (0..<coordinate.row).reversed() {
+                let currentCoordinate = Coordinate(row: row, col: coordinate.col)
+                guard let text = tetrisGrid.remove(at: currentCoordinate) else {
+                    break
+                }
+                let newCoordinate = currentCoordinate.getNextRow()
+                tetrisGrid.add(at: newCoordinate, tileText: text)
+                coordinatesToShift.append((from: currentCoordinate, to: newCoordinate))
+            }
+        }
+        delegate?.animate(shiftTiles: coordinatesToShift)
+    }
+
+    // TODO: fetch from database to match valid phrase
+    private func getValidPhraseRange(_ line: String) -> CountableRange<Int>? {
+        let phrases = ["先发制人"]
+        for phrase in phrases {
+            let phraseLen = phrase.characters.count
+            for i in 0...line.characters.count - phraseLen {
+                let startIndex = line.index(line.startIndex, offsetBy: i)
+                if line[startIndex..<line.index(startIndex, offsetBy: phraseLen)] == phrase {
+                    return i..<i + phraseLen
+                }
+            }
+        }
+        return nil
+    }
+
+    // TODO: generate from database, base on existing grid
+    fileprivate func getRandomWord() -> String {
+        let words = "先发制人"
+        return String(words[words.index(words.startIndex,
+                                        offsetBy: Random.integer(toInclusive: UInt(words.characters.count)))])
+    }
+}
+
+extension TetrisGameViewModel: TetrisGameViewModelProtocol {
+
+    func shiftFallingTile(to coordinate: Coordinate) -> Bool {
+        return !tetrisGrid.hasTile(at: coordinate)
+    }
+
+    func dropNextTile() -> (location: Coordinate, tileText: String) {
+        upcomingTiles.append(getRandomWord())
+        let tileText = upcomingTiles.removeFirst()
+        delegate?.redisplayUpcomingTiles()
+        fallingTileText = tileText
```

Comment:
cos `fallingTileText` is optional, so can't use it for the return type without unwrapping, then might as well just assign the result of `removeFirst` to a non-optional